src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -1,100 +1,290 @@
             {
-                if (_symbolCountByProjectMap != null)
+                Compilation compilation = null;
+                foreach (var reference in project.MetadataReferences.OfType<PortableExecutableReference>())
                 {
-                    _symbolCountByProjectMap.Remove(projectId);
+                    compilation = await UpdateReferenceAsync(project, reference, compilation, cancellationToken).ConfigureAwait(false);
                 }
             }
 
-            private void ReportCount()
+            private async Task<Compilation> UpdateReferenceAsync(
+                Project project, PortableExecutableReference reference, Compilation compilation, CancellationToken cancellationToken)
             {
-                var sourceSymbolCount = _symbolCountByProjectMap.Sum(kv => kv.Value);
-                Logger.Log(FunctionId.Run_Environment, KeyValueLogMessage.Create(m => m["SourceSymbolCount"] = sourceSymbolCount));
+                var key = GetReferenceKey(reference);
+                if (key != null)
+                {
+                    DateTime lastWriteTime;
+                    if (!TryGetLastWriteTime(key, out lastWriteTime))
+                    {
+                        // Couldn't get the write time.  Just ignore this reference.
+                        return compilation;
+                    }
+
+                    MetadataInfo metadataInfo;
+                    if (_metadataPathToInfo.TryGetValue(key, out metadataInfo) && metadataInfo.TimeStamp == lastWriteTime) 
+                    {
+                        // We've already computed and cached the info for this reference.
+                        return compilation;
+                    }
 
-                // we only report it once
-                _symbolCountByProjectMap = null;
-                _solutionId = null;
+                    compilation = compilation ?? await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                    var assembly = compilation.GetAssemblyOrModuleSymbol(reference) as IAssemblySymbol;
+                    if (assembly != null)
+                    {
+                        var info = await SymbolTreeInfo.TryGetInfoForMetadataAssemblyAsync(project.Solution, assembly, reference, loadOnly: false, cancellationToken: cancellationToken).ConfigureAwait(false);
+                        metadataInfo = metadataInfo ?? new MetadataInfo(lastWriteTime, info, new HashSet<ProjectId>());
+
+                        // Keep track that this dll is referenced by this project.
+                        metadataInfo.ReferencingProjects.Add(project.Id);
+
+                        _metadataPathToInfo.AddOrUpdate(key, metadataInfo, (_1, _2) => metadataInfo);
+                    }
+                }
+
+                return compilation;
+            }
+
+            public override void RemoveProject(ProjectId projectId)
+            {
+                ProjectInfo tuple;
```

Comment:
Does this get called when a solution is closed?  Is there a faster path that we can use that just `Clear()`s both dictionaries?